pkg/cli/mirror/catalog_images.go

```diff
@@ -233,6 +310,249 @@ func (o *MirrorOptions) processCatalogRefs(ctx context.Context, catalogsByImage
+// use upon rebuilding the catalog: This is because the opm binary can be called `opm` but also
+// `darwin-amd64-opm` etc.
+func extractOPMAndCache(ctx context.Context, srcRef image.TypedImageReference, outDir string, insecure bool) error {
+	var img v1.Image
+	var err error
+	refExact := srcRef.Ref.Exact()
+	if srcRef.OCIFBCPath == "" {
+		remoteOpts := getCraneOpts(ctx, insecure)
+		img, err = crane.Pull(refExact, remoteOpts...)
+		if err != nil {
+			return fmt.Errorf("unable to pull image from %s: %v", refExact, err)
+		}
+	} else {
+		img, err = getPtfImageFromOCIIndex(v1alpha2.TrimProtocol(srcRef.OCIFBCPath), runtime.GOARCH, runtime.GOOS)
+		if err != nil {
+			return err
+		}
+	}
+	// if we get here and no image was found bail out
+	if img == nil {
+		return fmt.Errorf("unable to obtain image for %v", srcRef)
+	}
+	cachePath, err := getCachePath(img)
+	if err != nil {
+		if errors.Is(err, NoCacheArgsError) {
+			return nil
+		} else {
+			return err
+		}
+	}
+
+	cacheLocationFileName := filepath.Join(outDir, cacheLocationPlaceholder)
+
+	baseDir := filepath.Dir(cacheLocationFileName)
+	err = os.MkdirAll(baseDir, 0755)
+	if err != nil {
+		return err
+	}
+
+	cfl, err := os.Create(cacheLocationFileName)
+	if err == nil {
+		defer cfl.Close()
+	} else {
+		return err
+	}
+
+	_, err = cfl.Write([]byte(cachePath))
+	if err != nil {
+		return err
+	}
+	cfl.Close()
+	// cachePath exists, opm binary will be needed to regenerate it
+	tr := tar.NewReader(mutate.Extract(img))
+	opmBinaryExtracted := ""
+	for {
+		header, err := tr.Next()
+
+		// break the infinite loop when EOF
+		if errors.Is(err, io.EOF) {
+			break
+		}
+
+		// skip the file if it is a directory or if file name does not end with `opm`
+		if !strings.HasSuffix(header.Name, opmBinarySuffix) || header.FileInfo().IsDir() {
+			continue
+		}
+
+		var buf bytes.Buffer
+		_, err = buf.ReadFrom(tr)
+		if err != nil {
+			return err
+		}
+
+		targetFileName := filepath.Join(outDir, config.OpmBinDir, header.Name)
+		bytes := buf.Bytes()
+
+		baseDir := filepath.Dir(targetFileName)
+		err = os.MkdirAll(baseDir, 0755)
+		if err != nil {
+			return err
+		}
+
+		f, err := os.Create(targetFileName)
+		if err == nil {
+			defer f.Close()
+		} else {
+			return err
+		}
+
+		_, err = f.Write(bytes)
+		if err != nil {
+			return err
+		}
+		opmBinaryExtracted = targetFileName
+
+		// check for the extracted opm file (it should exist if we found something)
+		_, err = os.Stat(opmBinaryExtracted)
+		if errors.Is(err, os.ErrNotExist) {
+			return fmt.Errorf("opm binary not found after extracting %q from catalog image %v", opmBinaryExtracted, srcRef)
+		}
+	}
+
+	return nil
+}
+
+// getPtfImageFromOCIIndex takes an oci local image located in `fbcPath` and finds the image that
+// corresponds to the current platform and OS within the manifestList or imageIndex
+func getPtfImageFromOCIIndex(fbcPath string, architecture string, os string) (v1.Image, error) {
```

Comment:
Nit: is `Ptf` in this function name supposed to be platform? Took a while for me to reason out what this meant.

Reply:
Sure . renamed